### PR TITLE
Support Simba JDBC driver for product-tests

### PR DESCRIPTION
**Pull request by brian-rickman**
The SQL (convention) based tests all pass using the open source JDBC driver or the Simba JDBC driver.  

2 tests in the JDBC group used methods which are only present in the open source JDBC driver, so those are skipped when the test is run with the Simba JDBC driver.  

1 test in the JDBC group fails when run with the Simba driver: `JdbcTests.shouldGetTableTypes`.  This failure indicates a bug which Simba will address shortly.

All tests in the SYSTEM, JMX, and HIVE_CONNECTOR groups now pass with either JDBC driver.

### CODE REVIEW

File: presto-product-tests/src/main/resources/sql-tests/testcases/catalog/describe.result [hunk 1]
```diff
@@ -1,4 +1,4 @@
--- delimiter: |; trimValues:true; types: LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/sql-tests/testcases/catalog/describe.result:1]: I like that! :)

File: presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q10.result [hunk 2]
```diff
@@ -1,4 +1,4 @@
--- delimiter: |; ignoreOrder: false; types: BIGINT|LONGNVARCHAR|DOUBLE|DOUBLE|LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q10.result:1]: what about q11 and other tpch tests?

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 3]
```diff
@@ -59,13 +58,13 @@ public Requirement getRequirements(Configuration configuration)
         }
     }
 
-    private PrestoConnection connection;
+    private java.sql.Connection connection;
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: use import

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 4]
```diff
@@ -101,8 +100,8 @@ public void shouldInsertSelectQuery()
     public void shouldExecuteQueryWithSelectedCatalogAndSchema()
             throws SQLException
     {
-        connection.setCatalog("hive");
-        connection.setSchema("default");
+       connection.setCatalog("hive");
+       connection.setSchema("default");
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: undo

Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: undo

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 5]
```diff
@@ -113,11 +112,13 @@ public void shouldExecuteQueryWithSelectedCatalogAndSchema()
     public void shouldSetTimezone()
             throws SQLException
     {
-        String timeZoneId = "Indian/Kerguelen";
-        connection.setTimeZoneId(timeZoneId);
-        try (Statement statement = connection.createStatement()) {
-            QueryResult result = queryResult(statement, "select current_timezone()");
-            assertThat(result).contains(row(timeZoneId));
+        if (connection instanceof PrestoConnection) {
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: otherwise test should not be enabled for simba. So instead of if please use Precondition.checkState()

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 6]
```diff
@@ -125,10 +126,12 @@ public void shouldSetTimezone()
     public void shouldSetLocale()
             throws SQLException
     {
-        connection.setLocale(CHINESE);
-        try (Statement statement = connection.createStatement()) {
-            QueryResult result = queryResult(statement, "SELECT date_format(TIMESTAMP '2001-01-09 09:04', '%M')");
-            assertThat(result).contains(row("一月"));
+        if (connection instanceof PrestoConnection) {
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: ditto

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 7]
```diff
@@ -173,9 +176,9 @@ private QueryResult queryResult(Statement statement, String query)
         return QueryResult.forResultSet(statement.executeQuery(query));
     }
 
-    private PrestoDatabaseMetaData metaData()
+    private java.sql.DatabaseMetaData metaData()
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: use import

File: presto-product-tests/src/main/resources/com/facebook/presto/tests/jdbc/get_nation_columns.result [hunk 8]
```diff
@@ -1,5 +1,5 @@
--- delimiter: |; ignoreOrder: false; ignoreExcessRows: false; types: LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR|BIGINT|LONGNVARCHAR|BIGINT|BIGINT|BIGINT|BIGINT|BIGINT|LONGNVARCHAR|LONGNVARCHAR|BIGINT|BIGINT|BIGINT|BIGINT|LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR|LONGNVARCHAR|BIGINT|LONGNVARCHAR|LONGNVARCHAR
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/com/facebook/presto/tests/jdbc/get_nation_columns.result:1]: I think this belongs to different commit

File: presto-product-tests/src/main/resources/table-results/presto-nation.result [hunk 9]
```diff
@@ -1,4 +1,4 @@
--- delimiter: |; ignoreOrder: true; types: BIGINT|LONGNVARCHAR|BIGINT|LONGNVARCHAR
```
Comment (kokosing) [inline presto-product-tests/src/main/resources/table-results/presto-nation.result:1]: ditto

File: presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java [hunk 10]
```diff
@@ -50,16 +61,25 @@ public void selectRuntimeQueries()
...
-                        BIGINT, BIGINT, BIGINT, TIMESTAMP, TIMESTAMP, TIMESTAMP, LONGNVARCHAR)
-                .hasAnyRows();
+                "FROM system.runtime.queries";
+        if (usingOpenSourceJdbcDriver()) {
+            assertThat(query(sql))
+                    .hasColumns(LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR,
+                            BIGINT, BIGINT, BIGINT, TIMESTAMP, TIMESTAMP, TIMESTAMP, LONGNVARCHAR)
+                    .hasAnyRows();
+        }
+        else {
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java]:
```
else usingSimbaDriver()  {
  ...
} else throw new InvalidStateException();
```

Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java]: please fix that in all the tests

File: presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java [hunk 11]
```diff
@@ -29,15 +32,23 @@
...
-                .hasAnyRows();
+        String sql = "SELECT node_id, http_uri, node_version, state FROM system.runtime.nodes";
+        if (usingOpenSourceJdbcDriver()) {
+            assertThat(query(sql))
+                    .hasColumns(LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR, LONGNVARCHAR)
+                    .hasAnyRows();
+        }
+        else {
+            assertThat(query(sql))
+                    .hasColumns(VARCHAR, VARCHAR, VARCHAR, VARCHAR)
```
Comment (kokosing) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java]: I wonder if it is a desired behaviour. Maybe should file a ticket to Simba? What do you think?

**Comment by kokosing** [comment]
I would squash all the commits together as they are related to the same issue.

I wonder what to do with such driver differences. Which behaviour is correct? Both? Generally it is not about making tests pass but to make sure that driver does what it is ought to do.

Can you please check what data types jdbc driver should return (in some JDBC standard documentation) for types where different type is returned by different driver?

Also what about session time zone? How JDBC standarize that?

File: presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q10.result (hunk 2 above)
Comment (brian-rickman) [inline presto-product-tests/src/main/resources/sql-tests/testcases/hive_tpch/q10.result:1]: I see the value of being consistent, but I was trying to make the smallest set of changes that would solve the problem.  The other tests don't use VARCHAR columns, and so can pass without this change.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java (hunk 5 above)
Comment (brian-rickman) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Preconditions.html#checkState(boolean)
That would throw an exception when it is run with Simba's driver, which would cause the test to fail .  Unless we put a try-catch block around it, but that seems a bit hacky to me.

File: presto-product-tests/src/main/resources/com/facebook/presto/tests/jdbc/get_nation_columns.result (hunk 8 above)
Comment (brian-rickman) [inline presto-product-tests/src/main/resources/com/facebook/presto/tests/jdbc/get_nation_columns.result:1]: I'll smoosh all the commits together before merging.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java (hunk 11 above)
Comment (brian-rickman) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/SystemConnectorTests.java]: I checked with Simba, online JDBC docs, and Presto CLI.  The actual column type in Presto is VARCHAR, so that is the correct value.  I filed [GitHub 4797](https://github.com/facebook/presto/issues/4797) because the open source driver is returning LONGNVARCHAR when it should be returning VARCHAR.

I could eliminate the if and let it fail with the open source driver, but I thought that might be annoying since we use the open source driver for the bulk of our tests.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 12]
```diff
@@ -100,8 +104,8 @@ public void shouldInsertSelectQuery()
     public void shouldExecuteQueryWithSelectedCatalogAndSchema()
             throws SQLException
     {
-       connection.setCatalog("hive");
-       connection.setSchema("default");
+        connection.setCatalog("hive");
+        connection.setSchema("default");
```
Comment (brian-rickman) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: This get the alignment back to where it was originally.

File: presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java [hunk 13]
```diff
@@ -112,7 +116,7 @@ public void shouldExecuteQueryWithSelectedCatalogAndSchema()
     public void shouldSetTimezone()
             throws SQLException
     {
-        if (connection instanceof PrestoConnection) {
+        if (usingFacebookJdbcDriver()) {
             String timeZoneId = "Indian/Kerguelen";
```
Comment (brian-rickman) [inline presto-product-tests/src/main/java/com/facebook/presto/tests/jdbc/JdbcTests.java]: Simba sets the timezone via a JDBC URL parameter, so there really isn't any equivalent.
